core/tee/tee_fs_key_manager.c

```diff
@@ -0,0 +1,430 @@
+		 *
+		 *     SSK = HMAC(HUK, message)
+		 *     message := concatenate(chip_id, static string)
+		 * */
+
+		message = malloc(sizeof(chip_id) + sizeof(string_for_ssk_gen));
+		if (!message) {
+			res = TEE_ERROR_OUT_OF_MEMORY;
+			goto exit;
+		}
+
+		tee_otp_get_hw_unique_key(&huk);
+		tee_otp_get_die_id(chip_id, sizeof(chip_id));
+
+		memcpy(message, chip_id, sizeof(chip_id));
+		memcpy(message + sizeof(chip_id), string_for_ssk_gen,
+				sizeof(string_for_ssk_gen));
+
+		res = generate_ssk(enc_fs_ssk.key, sizeof(enc_fs_ssk.key),
+				huk.data, sizeof(huk.data),
+				message, sizeof(message));
+
+		free(message);
+
+		if (res == TEE_SUCCESS)
+			enc_fs_ssk.is_init = 1;
+	}
+
+exit:
+	return res;
+}
+
+static TEE_Result do_auth_enc(TEE_OperationMode mode,
+		struct km_header *hdr,
+		uint8_t *fek, int fek_len,
+		const uint8_t *data_in, size_t in_size,
+		uint8_t *data_out, size_t *out_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t *ctx = NULL;
+	size_t ctx_size;
+	size_t tag_len = TEE_FS_KM_MAX_TAG_LEN;
+
+	if ((mode != TEE_MODE_ENCRYPT) && (mode != TEE_MODE_DECRYPT))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (in_size != *out_size)
+		return TEE_ERROR_SHORT_BUFFER;
+
+	res = crypto_ops.authenc.get_ctx_size(TEE_FS_KM_AUTH_ENC_ALG,
+			&ctx_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	ctx = malloc(ctx_size);
+	if (!ctx) {
+		EMSG("request memory size %zu failed", ctx_size);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	res = crypto_ops.authenc.init(ctx, TEE_FS_KM_AUTH_ENC_ALG,
+			mode, fek, fek_len, hdr->aad.iv,
+			TEE_FS_KM_IV_LEN, TEE_FS_KM_MAX_TAG_LEN,
+			sizeof(struct aad), in_size);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.update_aad(ctx, TEE_FS_KM_AUTH_ENC_ALG,
+			mode, (uint8_t *)hdr->aad.encrypted_key,
+			TEE_FS_KM_FEK_SIZE);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.update_aad(ctx, TEE_FS_KM_AUTH_ENC_ALG,
+			mode, (uint8_t *)hdr->aad.iv,
+			TEE_FS_KM_IV_LEN);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	if (mode == TEE_MODE_ENCRYPT) {
+		res = crypto_ops.authenc.enc_final(ctx, TEE_FS_KM_AUTH_ENC_ALG,
+				data_in, in_size, data_out, out_size,
+				hdr->tag, &tag_len);
+	} else {
+		res = crypto_ops.authenc.dec_final(ctx, TEE_FS_KM_AUTH_ENC_ALG,
+				data_in, in_size, data_out, out_size,
+				hdr->tag, tag_len);
+	}
+
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	crypto_ops.authenc.final(ctx, TEE_FS_KM_AUTH_ENC_ALG);
+
+exit:
+	free(ctx);
+	return res;
+}
+
+size_t tee_fs_get_header_size(enum tee_fs_file_type type)
+{
+	size_t header_size = 0;
+
+	switch (type) {
+	case META_FILE:
+		header_size = sizeof(struct meta_header);
+		break;
+	case BLOCK_FILE:
+		header_size = sizeof(struct block_header);
+		break;
+	default:
+		EMSG("Unknown file type, type=%d", type);
+		TEE_ASSERT(0);
+	}
+
+	return header_size;
+}
+
+TEE_Result tee_fs_generate_fek(uint8_t *encrypted_fek, int fek_size)
+{
+	TEE_Result res;
+
+	if (fek_size != TEE_FS_KM_FEK_SIZE)
```

Comment:
Why do we have to provide `fek_size` if we're only allowing a single size anyway?